stdenv: start deprecating non-list cmakeFlags

Artturin opened the pull request:
the motivation for this is to simplify stdenv and ease the job of
reviewers due to them needing to tell contributors about the defacto
rule that cmakeFlags should be a list of strings

###### Description of changes


continuation of https://github.com/NixOS/nixpkgs/pull/173172
this is why I made the deprecation release of them `23.05`

meson will be done in a pr after this one is merged

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Artturin committed:
stdenv: start deprecating non-list cmakeFlags

the motivation for this is to simplify stdenv and ease the job of
reviewers due to them needing to tell contributors about the defacto
rule that cmakeFlags should be a list of strings

Artturin committed:
treewide: convert string cmakeFlags to list of strings

Artturin commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -330,6 +330,25 @@ else let
+        (/**/ if lib.isString cmakeFlags then lib.warn "String 'cmakeFlags' is deprecated and will be removed in release 23.05. Please use a list of strings. Derivation name: ${derivationArg.name}, file: ${pos.file or "unknown file"}" [cmakeFlags]
+         else if cmakeFlags == null      then lib.warn "Null 'cmakeFlags' is deprecated and will be removed in release 23.05. Please use a empty list instead '[]'. Derivation name: ${derivationArg.name}, file: ${pos.file or "unknown file"}" []
+         else                                     cmakeFlags)
+        ++ optionals (stdenv.hostPlatform != stdenv.buildPlatform) ([ ]
+        ++ [ "-DCMAKE_SYSTEM_NAME=${lib.findFirst lib.isString "Generic" (lib.optional (!stdenv.hostPlatform.isRedox) stdenv.hostPlatform.uname.system)}"]
```
easier to read like this

SuperSandro2000 commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -330,6 +330,26 @@ else let
+        (/**/ if lib.isString cmakeFlags then lib.warn "String 'cmakeFlags' is deprecated and will be removed in release 23.05. Please use a list of strings. Derivation name: ${derivationArg.name}, file: ${pos.file or "unknown file"}" [cmakeFlags]
+         else if cmakeFlags == null      then lib.warn "Null 'cmakeFlags' is deprecated and will be removed in release 23.05. Please use a empty list instead '[]'. Derivation name: ${derivationArg.name}, file: ${pos.file or "unknown file"}" []
+         else                                     cmakeFlags)
+        ++ optionals (stdenv.hostPlatform != stdenv.buildPlatform) (
+          [ "-DCMAKE_SYSTEM_NAME=${lib.findFirst lib.isString "Generic" (lib.optional (!stdenv.hostPlatform.isRedox) stdenv.hostPlatform.uname.system)}"]
+        ++ optionals (stdenv.hostPlatform.uname.processor != null)
+          [ "-DCMAKE_SYSTEM_PROCESSOR=${stdenv.hostPlatform.uname.processor}" ]
+        ++ optionals (stdenv.hostPlatform.uname.release != null)
+          [ "-DCMAKE_SYSTEM_VERSION=${stdenv.hostPlatform.uname.release}" ]
+        ++ optionals (stdenv.hostPlatform.isDarwin)
+          [ "-DCMAKE_OSX_ARCHITECTURES=${stdenv.hostPlatform.darwinArch}" ]
+        ++ optionals (stdenv.buildPlatform.uname.system != null)
+          [ "-DCMAKE_HOST_SYSTEM_NAME=${stdenv.buildPlatform.uname.system}" ]
+        ++ optionals (stdenv.buildPlatform.uname.processor != null)
+          [ "-DCMAKE_HOST_SYSTEM_PROCESSOR=${stdenv.buildPlatform.uname.processor}" ]
+        ++ optionals (stdenv.buildPlatform.uname.release != null)
+          [ "-DCMAKE_HOST_SYSTEM_VERSION=${stdenv.buildPlatform.uname.release}" ]
```
I actually liked the more compact form it had before more

Artturin replied:
switched it to be more compact

magnetophon reviewed (approved):
At least the faust part LGTM.

Artturin commented:
```
$ gco depcmakestring
Switched to branch 'depcmakestring'
Your branch is up to date with 'origin/depcmakestring'.
$ nix eval ".#pkgsCross.aarch64-multiplatform.h3.cmakeFlags"
[ "-DBUILD_SHARED_LIBS=ON" "-DENABLE_LINTING=OFF" "-DCMAKE_SYSTEM_NAME=Linux" "-DCMAKE_SYSTEM_PROCESSOR=aarch64" "-DCMAKE_HOST_SYSTEM_NAME=Linux" "-DCMAKE_HOST_SYSTEM_PROCESSOR=x86_64" ]
$ nix eval ".#h3.cmakeFlags"
[ "-DBUILD_SHARED_LIBS=ON" "-DENABLE_LINTING=OFF" ]
$ gcm
Switched to branch 'master'
Your branch is up to date with 'upstream/master'.
$ nix eval ".#pkgsCross.aarch64-multiplatform.h3.cmakeFlags"
[ "-DBUILD_SHARED_LIBS=ON" "-DENABLE_LINTING=OFF" "-DCMAKE_SYSTEM_NAME=Linux" "-DCMAKE_SYSTEM_PROCESSOR=aarch64" "-DCMAKE_HOST_SYSTEM_NAME=Linux" "-DCMAKE_HOST_SYSTEM_PROCESSOR=x86_64" ]
$ nix eval ".#h3.cmakeFlags"
[ "-DBUILD_SHARED_LIBS=ON" "-DENABLE_LINTING=OFF" ]
```

magnetophon reviewed (approved):
Not sure why my review was requested again, but I'll appease ofborg.

magnetophon commented:
I'm being stalked by ofborg!  ;)

roberth commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -330,6 +330,18 @@ else let
+         else if cmakeFlags == null      then lib.warn "Null 'cmakeFlags' is deprecated and will be removed in release 23.05. Please use a empty list instead '[]'. Derivation name: ${derivationArg.name}, file: ${pos.file or "unknown file"}" []
+         else                                 cmakeFlags)
+        ++ optionals (stdenv.hostPlatform != stdenv.buildPlatform) ([ "-DCMAKE_SYSTEM_NAME=${lib.findFirst lib.isString "Generic" (lib.optional (!stdenv.hostPlatform.isRedox) stdenv.hostPlatform.uname.system)}"]
+        ++ optionals (stdenv.hostPlatform.uname.processor != null)  [ "-DCMAKE_SYSTEM_PROCESSOR=${stdenv.hostPlatform.uname.processor}" ]
```
These inner `optionals` seem like they should be indented.

Maybe pull the entire `cmakeFlags` attribute value through `nixpkgs-fmt`? Now seems like the right time, because we're "losing" git blame anyway for this chunk.

```suggestion
            ++ optionals (stdenv.hostPlatform.uname.processor != null)  [ "-DCMAKE_SYSTEM_PROCESSOR=${stdenv.hostPlatform.uname.processor}" ]
```

Artturin replied:
Done

Mindavi commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -342,19 +356,6 @@ else let
       outputHashAlgo = attrs.outputHashAlgo or "sha256";
       outputHashMode = attrs.outputHashMode or "recursive";
     } // lib.optionalAttrs (stdenv.hostPlatform != stdenv.buildPlatform) {
-      cmakeFlags =
```
Heh, in this diff it looks like it was previously only applied when cross-compiling?

Artturin replied:
Yes

jtojnar commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -330,6 +330,20 @@ else let
+        (/**/ if lib.isString cmakeFlags then lib.warn "String 'cmakeFlags' is deprecated and will be removed in release 23.05. Please use a list of strings. Derivation name: ${derivationArg.name}, file: ${pos.file or "unknown file"}" [cmakeFlags]
+         else if cmakeFlags == null      then lib.warn "Null 'cmakeFlags' is deprecated and will be removed in release 23.05. Please use a empty list instead '[]'. Derivation name: ${derivationArg.name}, file: ${pos.file or "unknown file"}" []
+         else                                 cmakeFlags)
+          ++ optionals (stdenv.hostPlatform != stdenv.buildPlatform) (
```
The indentation here is weird – almost looks like it belongs to the else branch.

jtojnar replied:
Maybe something like the following to get rid of the alignment, which looks ugly with proportional fonts:

![blagh](https://user-images.githubusercontent.com/705123/177220540-cd36a14b-3592-4d4a-ae50-cb59931619f4.png)


```nix
      cmakeFlags =
        let
          inherit (lib) optionals;
          explicitFlags =
            if lib.isString cmakeFlags then
              lib.warn
                "String 'cmakeFlags' is deprecated and will be removed in release 23.05. Please use a list of strings. Derivation name: ${derivationArg.name}, file: ${pos.file or "unknown file"}"
                [cmakeFlags]
            else if cmakeFlags == null then
              lib.warn
                "Null 'cmakeFlags' is deprecated and will be removed in release 23.05. Please use a empty list instead '[]'. Derivation name: ${derivationArg.name}, file: ${pos.file or "unknown file"}"
                []
            else
              cmakeFlags;

          crossFlags = [
            "-DCMAKE_SYSTEM_NAME=${lib.findFirst lib.isString "Generic" (lib.optional (!stdenv.hostPlatform.isRedox) stdenv.hostPlatform.uname.system)}"
          ] ++ optionals (stdenv.hostPlatform.uname.processor != null) [
            "-DCMAKE_SYSTEM_PROCESSOR=${stdenv.hostPlatform.uname.processor}"
          ] ++ optionals (stdenv.hostPlatform.uname.release != null) [
            "-DCMAKE_SYSTEM_VERSION=${stdenv.hostPlatform.uname.release}"
          ] ++ optionals (stdenv.hostPlatform.isDarwin) [
            "-DCMAKE_OSX_ARCHITECTURES=${stdenv.hostPlatform.darwinArch}"
          ] ++ optionals (stdenv.buildPlatform.uname.system != null) [
            "-DCMAKE_HOST_SYSTEM_NAME=${stdenv.buildPlatform.uname.system}"
          ] ++ optionals (stdenv.buildPlatform.uname.processor != null) [
            "-DCMAKE_HOST_SYSTEM_PROCESSOR=${stdenv.buildPlatform.uname.processor}"
          ] ++ optionals (stdenv.buildPlatform.uname.release != null) [
            "-DCMAKE_HOST_SYSTEM_VERSION=${stdenv.buildPlatform.uname.release}"
          ];
        in
          explicitFlags
          ++ optionals (stdenv.hostPlatform != stdenv.buildPlatform) crossFlags;
```

Artturin replied:
The lists were on the different lines before Sandro suggested they be on the same line

Also luckily everyone uses monospace fonts for programming

ckiee replied:
> Also luckily everyone uses monospace fonts for programming

There's a good chance that they aren't if they noticed it :P

Artturin replied:
applied the suggestion but reduced the vertical space need

SuperSandro2000 replied:
> Also luckily everyone uses monospace fonts for programming

yes, JetBrains Mono

I don't like really like the code below. Its just a giant block and before we could easily see what happens but we bikesheeded enough here.

```nix
          crossFlags = [
            "-DCMAKE_SYSTEM_NAME=${lib.findFirst lib.isString "Generic" (lib.optional (!stdenv.hostPlatform.isRedox) stdenv.hostPlatform.uname.system)}"
          ] ++ optionals (stdenv.hostPlatform.uname.processor != null) [
            "-DCMAKE_SYSTEM_PROCESSOR=${stdenv.hostPlatform.uname.processor}"
          ] ++ optionals (stdenv.hostPlatform.uname.release != null) [
            "-DCMAKE_SYSTEM_VERSION=${stdenv.hostPlatform.uname.release}"
          ] ++ optionals (stdenv.hostPlatform.isDarwin) [
            "-DCMAKE_OSX_ARCHITECTURES=${stdenv.hostPlatform.darwinArch}"
          ] ++ optionals (stdenv.buildPlatform.uname.system != null) [
            "-DCMAKE_HOST_SYSTEM_NAME=${stdenv.buildPlatform.uname.system}"
          ] ++ optionals (stdenv.buildPlatform.uname.processor != null) [
            "-DCMAKE_HOST_SYSTEM_PROCESSOR=${stdenv.buildPlatform.uname.processor}"
          ] ++ optionals (stdenv.buildPlatform.uname.release != null) [
            "-DCMAKE_HOST_SYSTEM_VERSION=${stdenv.buildPlatform.uname.release}"
          ];
```

Artturin commented:
no more style changed please :P

Artturin commented on pkgs/applications/audio/faust/faust2.nix:
```diff
@@ -58,9 +58,10 @@ let
+    cmakeFlags = [
+      "-C../backends/all.cmake"
+      "-C../targets/all.cmake"
+    ];
```
https://github.com/NixOS/nixpkgs/pull/78613#discussion_r912571372

Artturin commented:
https://github.com/NixOS/nixpkgs/pull/180302
